src/darwin/Framework/CHIP/MTRSetupPayload.h

```diff
@@ -116,55 +181,81 @@ MTR_AVAILABLE(ios(16.1), macos(13.0), watchos(9.1), tvos(16.1))
- */
-+ (MTRSetupPayload * _Nullable)setupPayloadWithOnboardingPayload:(NSString *)onboardingPayload
-                                                           error:(NSError * __autoreleasing *)error
-    MTR_AVAILABLE(ios(16.2), macos(13.1), watchos(9.2), tvos(16.2));
-
 /**
  * Initialize an MTRSetupPayload with the given passcode and discriminator.
  * This will pre-set version, product id, and vendor id to 0.
  */
 - (instancetype)initWithSetupPasscode:(NSNumber *)setupPasscode
                         discriminator:(NSNumber *)discriminator MTR_AVAILABLE(ios(16.2), macos(13.1), watchos(9.2), tvos(16.2));
 
-/** Get 11 digit manual entry code from the setup payload. */
-- (NSString * _Nullable)manualEntryCode;
+/**
+ * Creates a Manual Pairing Code from this setup payload.
+ * Returns nil if this payload cannot be represented as a valid Manual Pairing Code.
+ *
+ * The following properties must be populated for a valid Manual Pairing Code:
+ *  - setupPasscode
+ *  - discriminator (short or long)
+ *
+ * In most cases the pairing code will be 11 digits long. If the payload indicates
+ * a `commissioningFlow` other than `MTRCommissioningFlowStandard`, a 21 digit code
+ * will be produced that includes the vendorID and productID values.
+ */
+- (nullable NSString *)manualEntryCode;
 
 /**
- * Get a QR code from the setup payload.
+ * Creates a QR Code payload from this setup payload.
+ * Returns nil if this payload cannot be represented as a valid QR Code.
  *
- * Returns nil on failure (e.g. if the setup payload does not have all the
- * information a QR code needs).
+ * The following properties must be populated for a valid QR Code:
+ * - setupPasscode
+ * - discriminator (must be long)
+ * - discoveryCapabilities (not MTRDiscoveryCapabilitiesUnknown)
  */
-- (NSString * _Nullable)qrCodeString:(NSError * __autoreleasing *)error
-    MTR_AVAILABLE(ios(16.2), macos(13.1), watchos(9.2), tvos(16.2));
+- (NSString * _Nullable)qrCodeString MTR_NEWLY_AVAILABLE;
 
 @end
 
 @interface MTROptionalQRCodeInfo (Deprecated)
 
-@property (nonatomic, copy)
-    NSNumber * infoType MTR_DEPRECATED("Please use type", ios(16.1, 16.4), macos(13.0, 13.3), watchos(9.1, 9.4), tvos(16.1, 16.4));
+- (instancetype)init MTR_NEWLY_DEPRECATED("Please use -initWithTag:...value:");
+
+@property (nonatomic, copy) NSNumber * infoType
+    MTR_DEPRECATED("Please use type", ios(16.1, 16.4), macos(13.0, 13.3), watchos(9.1, 9.4), tvos(16.1, 16.4));
 
+@property (nonatomic, copy) NSNumber * tag MTR_NEWLY_DEPRECATED("Please use tagNumber");
+
+- (void)setType:(MTROptionalQRCodeInfoType)type MTR_NEWLY_DEPRECATED("MTROptionalQRCodeInfo is immutable");
+- (void)setTag:(NSNumber *)tag MTR_NEWLY_DEPRECATED("MTROptionalQRCodeInfo is immutable");
+- (void)setIntegerValue:(NSNumber *)integerValue MTR_NEWLY_DEPRECATED("MTROptionalQRCodeInfo is immutable");
+- (void)setStringValue:(NSString *)stringValue MTR_NEWLY_DEPRECATED("MTROptionalQRCodeInfo is immutable");
```

Comment:
I’ll have to check that

Reply:
So, it's ABI compatible because we're using the ObjC ABI, but assigning to them is not source compatible.

That's of course conceptually not great, however I don't think there was so far any practical reason to actually assign to these properties, because we had no API for adding a vendor element. The reason I think these should be immutable is that exposing a list of mutable objects gives the impression that these can be mutated in-place.

Not sure what level of effort we want to invest to fix this since it's in practice likely a non-issue. We could deprecated `MTROptionalQRCodeInfo` and make it sub-class a newly introduced `MTRSetupPayloadVendorElement` (or similar) and have the properties be mutable in the sub-class only, and the new APIs would use the new class. But I don't know if this is worth the trouble.

Reply:
LGTM